compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -771,6 +776,9 @@ object Parsers {
         }
         else infixType()
 
+      if (isImplicit && !isImplicitFun && in.token != ARROW)
+        syntaxError("Types with implicit keyword can only be function types", start, nme.IMPLICITkw)
+
       in.token match {
         case ARROW => functionRest(t :: Nil)
         case FORSOME => syntaxError(ExistentialTypesNoLongerSupported()); t
```

Comment:
I think you can avoid the `isImplicitFun` logic and simply change the next line to:

     case _ => 
        if (isImplicit) syntaxError("Types with implicit keyword can only be function types", start)
        t